sw/device/sca/sha3_serial.c

```diff
@@ -471,6 +500,52 @@ static void sha3_serial_single_absorb(const uint8_t *msg, size_t msg_len) {
   kmac_reset();
 }
 
+static void sha3_serial_fixed_key_set(const uint8_t *key, size_t key_len) {
+  SS_CHECK(key_len == kKeyLength);
+  memcpy(key_fixed, key, key_len);
+}
+
+static void sha3_serial_batch(const uint8_t *data, size_t data_len) {
+  uint32_t num_encryptions = 0;
+  uint32_t out[kDigestLength];
+  uint32_t batch_digest[kDigestLength];
+  SS_CHECK(data_len == sizeof(num_encryptions));
+  num_encryptions = read_32(data);
+
+  // This bit may not be needed. Test afterwards
+  for (uint32_t j = 0; j < kDigestLength; ++j) {
+    batch_digest[j] = 0;
+  }
+
+  for (uint32_t i = 0; i < num_encryptions; ++i) {
+    if (run_fixed) {
+      memcpy(batch_keys[i], key_fixed, kKeyLength);
+    } else {
+      prng_rand_bytes(batch_keys[i], kKeyLength);
+    }
+    prng_rand_bytes(batch_plaintexts[i], kMessageLength);
+    run_fixed = batch_plaintexts[i][0] & 0x1;
+  }
+
+  for (uint32_t i = 0; i < num_encryptions; ++i) {
+    kmac_reset();
+    memcpy(kmac_key.share0, batch_keys[i], kKeyLength);
+
+    sca_set_trigger_high();
+    sha3_serial_absorb(batch_plaintexts[i], kMessageLength);
+    sca_set_trigger_low();
+
+    kmac_msg_done();
+    SS_CHECK_DIF_OK(kmac_get_digest(out, kDigestLength));
+
+    for (uint32_t j = 0; j < kDigestLength; ++j) {
```

Comment:
Do you want to add a comment saying that we basically chain all messages here? In terms of performance, @abdullahvarici figured out that for AES this chaining does have a very big impact on performance. Did you see something similar for KMAC as well?

Reply:
I agree. If we only read and compare a word of the last ciphertext of every batch, the performance would be better.

Reply:
The performance penalty doesn't seem to be very high for KMAC, possibly due to the fewer traces per batch. The capture speed is around 620 traces/s.
I've added the comment in the code explaining that we verify each batch by XORing all outputs in the batch and comparing the result. I would like to keep it as it is for now.

Reply:
This sounds good to me, thanks for summarizing the conclusion here and for making the comment in the code.